src/riverroute/mosart_driver.F90

```diff
@@ -0,0 +1,927 @@
+
+   ! budget
+   type(budget_type), public :: budget  ! type containing vars and routines for budget checking
+
+   character(len=CL) :: nlfilename_rof = 'mosart_in'
+   character(len=CL) :: fnamer              ! name of netcdf restart file
+
+   integer :: nt_liq, nt_ice
+
+   character(*), parameter :: u_FILE_u = &
+        __FILE__
+   !-----------------------------------------------------------------------
+
+contains
+
+   !-----------------------------------------------------------------------
+   subroutine mosart_read_namelist()
+      !
+      ! Read and distribute mosart namelist
+      !
+      ! local variables
+      integer           :: i
+      integer           :: ier       ! error code
+      integer           :: unitn     ! unit for namelist file
+      logical           :: lexist    ! File exists
+      character(len=CS) :: runtyp(4) ! run type
+      logical, allocatable :: do_euler_calc(:) ! turn on euler algorithm
+      character(len=*),parameter :: subname = '(mosart_read_namelist) '
+      !-----------------------------------------------------------------------
+
+      !-------------------------------------------------------
+      ! Read in mosart namelist
+      !-------------------------------------------------------
+
+      namelist /mosart_inparm / frivinp, finidat, nrevsn, coupling_period, ice_runoff, &
+           ndens, mfilt, nhtfrq, fincl1,  fincl2, fincl3, fexcl1,  fexcl2, fexcl3, &
+           avgflag_pertape, decomp_option, bypass_routing_option, qgwl_runoff_option, &
+           use_halo_option, delt_mosart, mosart_tracers, mosart_euler_calc, budget_frq, &
+           separate_glc2ocn_fluxes
+
+      ! Preset values
+      ice_runoff  = .true.
+      finidat = ' '
+      nrevsn  = ' '
+      coupling_period   = -1
+      delt_mosart = 3600
+      decomp_option = 'basin'
+      bypass_routing_option = 'direct_in_place'
+      qgwl_runoff_option = 'threshold'
+      use_halo_option = .false.
+      mosart_tracers = 'LIQ:ICE'
+      mosart_euler_calc = 'T:F'
+      separate_glc2ocn_fluxes = .false.
+
+      nlfilename_rof = "mosart_in" // trim(inst_suffix)
+      inquire (file = trim(nlfilename_rof), exist = lexist)
+      if ( .not. lexist ) then
+         write(iulog,*) subname // ' ERROR: nlfilename_rof does NOT exist: '//trim(nlfilename_rof)
+         call shr_sys_abort(trim(subname)//' ERROR nlfilename_rof does not exist')
+      end if
+      if (mainproc) then
+         write(iulog,*) 'Reading mosart_inparm namelist from: ', trim(nlfilename_rof)
+         open( newunit=unitn, file=trim(nlfilename_rof), status='old' )
+         ier = 1
+         do while ( ier /= 0 )
+            read(unitn, mosart_inparm, iostat=ier)
+            if (ier < 0) then
+               call shr_sys_abort( subname//' encountered end-of-file on mosart_inparm read' )
+            endif
+         end do
+         close(unitn)
+      end if
+
+      call mpi_bcast (finidat, len(finidat), MPI_CHARACTER, 0, mpicom_rof, ier)
```

Comment:
Originally I thought this was a good idea. Looking at this some more I found that ESMF_VMBroadcast only accepts arrays. That means that an array of length 1 needs to be created and then a copy of finidat needs to be made to that and then broadcast and then copied back. I think this will make the code more complicated and harder to read. 
At this point I recommend staying with native mpi calls. I'm happy to talk about this some more.

Reply:
See issue https://github.com/ESCOMP/MOSART/issues/100.